File: common/src/main/java/org/bread_experts_group/breadlib/registry/objects/RegistryBlock.java
package org.bread_experts_group.breadlib.registry.objects;

import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import org.bread_experts_group.breadlib.registry.ItemLikeExtended;
import org.jetbrains.annotations.NotNull;

public class RegistryBlock<B extends Block> extends RegistryObject<Block, B> implements ItemLikeExtended {
    public static <B extends Block> RegistryBlock<B> create(String modID, String name) {
        return new RegistryBlock<>(ResourceLocation.fromNamespaceAndPath(modID, name));
    }

    public RegistryBlock(ResourceLocation name) {
        super(name, BuiltInRegistries.BLOCK);
    }

    public BlockState defaultState() {
        return get().defaultBlockState();
    }

    @Override
    public @NotNull Item asItem() {
        return toStack().getItem();
    }

    @Override
    public ItemStack toStack() {
        return new ItemStack(this, 1);
    }

    @Override
    public ItemStack asStack(int count) {
        return new ItemStack(this, count);
    }
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/registry/objects/RegistryBlock.kt
package org.bread_experts_group.breadlib.registry.objects

import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import org.bread_experts_group.breadlib.registry.RegistryProvider.Companion.getItems

class RegistryBlock<B : Block>(name: ResourceLocation) : AbstractRegistryBlock<B>(
	name
) {
	companion object {
		fun <B : Block> create(modID: String, name: String): RegistryBlock<B> =
			RegistryBlock(ResourceLocation.fromNamespaceAndPath(modID, name))
	}

	fun withItem(properties: Item.Properties = Item.Properties()): RegistryBlockItem<B, BlockItem> {
		val blockItem = getItems(this.name.namespace).registerSimpleBlockItem(name.path, { this.get() }, properties)
		return RegistryBlockItem(this, blockItem)
	}
}
